docs/tasks.md

```diff
@@ -14,4 +14,37 @@ Examples of both `AddTask` calls can be found in the advection example [here](..
 `DoAvailable` loops over the task list once, executing all tasks whose dependencies are satisfied.  The function returns either `TaskListStatus::complete` if all tasks have been executed (and the task list is therefore empty) or `TaskListStatus::running` if tasks remain to be completed.
 
 ## TaskID
-The `TaskID` class implements methods that allow Parthenon to keep track of tasks, their dependencies, and what remains to be completed.  The main way application code will interact with this object is as a returned object from `TaskList::AddTask` and as an argument to subsequent calls to `TaskList::AddTask` as a dependency for other tasks.  When used as a dependency, `TaskID` objects can be combined with the bitwise or operator (`|`) to specify multiple dependencies.
\ No newline at end of file
+The `TaskID` class implements methods that allow Parthenon to keep track of tasks, their dependencies, and what remains to be completed.  The main way application code will interact with this object is as a returned object from `TaskList::AddTask` and as an argument to subsequent calls to `TaskList::AddTask` as a dependency for other tasks.  When used as a dependency, `TaskID` objects can be combined with the bitwise or operator (`|`) to specify multiple dependencies.
+
+ ## TaskRegion
+ `TaskRegion` is defined via a simple using statement as a `std::vector<TaskList>`.  During task execution (described below), all task lists in a `TaskRegion` can be operated on concurrently.  For example, a `TaskRegion` can be used to construct independent task lists for each `MeshBlock`.
+
+ ## TaskCollection
+ A `TaskCollection` contains a `std::vector<TaskRegion>`, i.e. an ordered list of `TaskRegion`s.  Importantly, each `TaskRegion` will be executed to completion before subsequent `TaskRegion`s, introducing a notion of sequential execution and enabling flexibility in task granularity.  For example, the following code fragment uses the `TaskCollection` and `TaskRegion` abstractions to express work that can be done asynchronously across blocks, followed by a bulk synchronous task involving all blocks, and finally another round of asynchronous work.
+ ```c++
+ TaskCollection tc;
+ TaskRegion &tr1 = tc.AddRegion(nmb);
+ for (int i = 0; i < nmb; i++) {
+   auto task_id = tr1[i].AddTask(foo, dep, args, blocks[i]);
+ }
+
+ {
+   TaskRegion &tr2 = tc.AddRegion(1);
+   auto sync_task = tr2[0].AddTask(bar, dep, args, blocks);
+ }
+
+ TaskRegion &tr3 = tc.AddRegion(nmb);
+ for (int i = 0; i < nmb; i++) {
+     auto task_id = tr3[i].AddTask(foo, dep, args, blocks[i]);
+ }
+ ```
+
+ `TaskCollection` provides two member functions, `AddRegion` and `Execute`.
+
```

Comment:
This description could really benefit from a diagram, illustrating the different components. Though, it would be more appropriate to include something like that in the wiki. Including something like that might be wishful thinking on my part though. I'm not quite sure we have a good way of including images related to documentation at this point.

Reply:
What kind of diagram are you thinking, @JoshuaSBrown ? It might be possible to draw one in ascii

Reply:
I guess I'm not sure what picture I would draw even if it was easy to include a picture.

Reply:
I don't know if this is accurate...

<img width="688" alt="TaskDiagram" src="https://user-images.githubusercontent.com/4664476/91745690-9e243f80-eb78-11ea-960f-8be1b115b33e.png">

Reply:
@JoshuaSBrown whoa nice work throwing that together.

Am I reading your picture correctly that the green blocks are tasks? And if they're one after the other, it's synchronous and if they're next to each other it's asynchronous? That's not inaccurate, I think. It doesn't really capture that tasks can be over one meshblock or many, and that if they're over one meshblock in a synchronous region, each meshblock likely individually owns the task. Not sure how to get that accross in a picture. Maybe we don't have to.

Regarding embedding a picture in the docs... I don't know if we want to do this, but we could just stick the picture in the `docs` directory and link to it with markdown `![name](path)`, which will make it appear embedded in github at least.

Could also try throwing it through a png->ascii converger. I think your graph is simple enough it should be representable as ascii

Reply:
@JoshuaSBrown can we iterate on this some and add it in a subsequent PR?

Reply:
That's fine with me.